### PR TITLE
Add functionality for metadata to provide some basic values

### PR DESCRIPTION
**Pull request by rajansandeep**
### 1. Why is this pull request needed and what does it do?
Extend Metadata functionality to include some basic data straight from the Metadata plugin.

### CODE REVIEW

**Comment by rajansandeep** [comment]
@fturib PTAL if the base is ok.
I have to extend the unit tests further.

File: plugin/metadata/request.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	QueryType  = "qtype"
+	ClientIP   = "client_ip"
+	ClientPort = "client_port"
+	Protocol   = "protocol"
+	ServerIP   = "server_ip"
+	ServerPort = "server_port"
+	ResponseIP = "response_ip"
+)
+
+func newRequest() *Metadata {
```
[the remark this answers is not in the document]
Comment (rajansandeep) [reply plugin/metadata/request.go]: okay, sgtm. Should I have a better name for the filename itself?

[the remark this answers is not in the document]
Comment (fturib) [reply plugin/metadata/request.go]: I think filename is ok : you are extracting data from the request to translate into metadata

**Comment by fturib** [comment]
I am ok with the changes.

and now we need the global test for usage: Metadata + Rewrite (Maybe in testt/test_rewrite ?)

File: test/metadata_test.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	r, err := dns.Exchange(m, server)
+	if err != nil {
+		t.Fatalf("Expected to receive reply, but didn't: %s", err)
+	}
+
+	// expect answer section with A record in it
+	if len(r.Answer) == 0 {
+		t.Error("Expected to at least one RR in the answer section, got none")
+	}
+	if r.Answer[0].Header().Rrtype != dns.TypeA {
```
Comment (fturib) [inline test/metadata_test.go]: why do you want to test if your query is type A ?

File: test/metadata_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		t.Errorf("Expected RR to A, got: %d", r.Answer[0].Header().Rrtype)
+	}
+	if r.Answer[0].(*dns.A).A.String() != "192.0.2.53" {
```
Comment (fturib) [inline test/metadata_test.go]: Why testing the IP address of example.com ? why if it changes ? does it change something for what you want to test ?

File: test/metadata_test.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	o := r.IsEdns0()
+	if o == nil || len(o.Option) == 0 {
+		t.Error("Expected EDNS0 options but got none")
+	} else {
+		if e, ok := o.Option[0].(*dns.EDNS0_LOCAL); ok {
+			if e.Code != 0xffee {
+				t.Errorf("Expected EDNS_LOCAL code 0xffee but got %x", e.Code)
+			}
+			if !bytes.Equal(e.Data, []byte("example.com.")) {
+				t.Errorf("Expected EDNS_LOCAL data 'example.com.' but got %q", e.Data)
```
Comment (fturib) [inline test/metadata_test.go]: YES. That is what you want to test here : that your metadata is really in the EDNS0 part

**Pull request closed without merging**